File: textmode.c
#include "textmode.h"

txmbuf earlytxm;

static void txm_scroll(txmbuf *buf, size_t line_count)
{
	if (line_count == 0)
		return;

	if (line_count >= GPUBUF_SCREEN_HEIGHT) {
		txm_clear_screen(buf);
		return;
	}

	for (int y = line_count; y < GPUBUF_SCREEN_HEIGHT; ++y) {
		for (u32 x = 0; x < GPUBUF_SCREEN_WIDTH; ++x) {
			(*buf->mem)[y-line_count][x] = (*buf->mem)[y][x];
		}
	}

	txm_clear_cur_line(buf);
}

void txm_putc(txmbuf *buf, char ch)
{
	(*buf->mem)[buf->cury][buf->curx].ch = ch;
	(*buf->mem)[buf->cury][buf->curx].bg = GPUBUF_CHAR_BG_BLACK;

	++buf->curx;
	if (buf->curx >= GPUBUF_SCREEN_WIDTH) {
		buf->curx = 0;
		++buf->cury;
		if (buf->cury >= GPUBUF_SCREEN_HEIGHT) {
			buf->cury = GPUBUF_SCREEN_HEIGHT - 1;
			txm_scroll(buf, 1);
		}
	}
}

void txm_clear_cur_line(txmbuf *buf)
{
	for (int i = 0; i < GPUBUF_SCREEN_WIDTH; ++i) {
		(*buf->mem)[buf->cury][i].ch = ' ';
		(*buf->mem)[buf->cury][i].bg = GPUBUF_CHAR_BG_BLACK;
	}
}

void txm_clear_screen(txmbuf *buf)
{
	for (int y = 0; y < GPUBUF_SCREEN_HEIGHT; ++y) {
		for (int x = 0; x < GPUBUF_SCREEN_WIDTH; ++x) {
			(*buf->mem)[y][x].ch = ' ';
			(*buf->mem)[y][x].bg = GPUBUF_CHAR_BG_BLACK;
		}
	}
}

void txm_print(txmbuf *buf, char *s)
{
	while (*s != 0) {
		txm_putc(buf, *s);
		++s;
	}
}

void txm_print_hex_u8(txmbuf *buf, u8 v)
{
	char digits[] = "0123456789ABCDEF";

	txm_putc(buf, digits[v >> 4]);
	txm_putc(buf, digits[v & 0x0f]);
}

void txm_print_hex(txmbuf *buf, u64 v)
{
	char digits[] = "0123456789ABCDEF";

	for (int i = 0; i < 16; ++i) {
		txm_putc(buf, digits[v >> 60]);
		v <<= 4;
	}
}

void txm_line_feed(txmbuf *buf)
{
	buf->curx = 0;
	++buf->cury;

	if (buf->cury >= GPUBUF_SCREEN_HEIGHT) {
		buf->cury = GPUBUF_SCREEN_HEIGHT - 1;
		txm_scroll(buf, 1);
	}
}

txmbuf make_early_txmbuf(void)
{
	txmbuf ret;
	ret.mem = (void*) GPUBUF;
	ret.cury = 0;
	ret.curx = 0;

	return ret;
}

File: kernel/src/textmode.rs
use core::ptr::write_volatile;

// This should probably go out of TextMode module at some point
pub trait OutputBuffer {
    fn putc(&mut self, c: char);
    fn puts(&mut self, s: &str);
    fn line_feed(&mut self);
}

trait TextModeBuffer {
    fn clear_cur_line(&mut self);
}

const GPUBUF               :u64   = 0xb8000;
const GPUBUF_SCREEN_WIDTH  :usize = 80;
const GPUBUF_SCREEN_HEIGHT :usize = 25;
const GPUBUF_CHAR_BG_BLACK :u8    = 0x0f;

#[repr(packed)]

// Disbale warning that the field is never read.
// This type is a view over VGA text mode buffer.
#[allow(dead_code)]

pub struct TxmColor {
    ch: u8,
    bg: u8,
}

pub struct TxmBuf {
    buf: &'static mut [[TxmColor; GPUBUF_SCREEN_WIDTH]; GPUBUF_SCREEN_HEIGHT],
    curx: usize,
    cury: usize,
}

impl TextModeBuffer for TxmBuf {
    fn clear_cur_line(&mut self) {
        for (i, c) in self.buf[self.cury].iter_mut().enumerate() {

            if i >= GPUBUF_SCREEN_WIDTH {
                break;
            }

            unsafe {
                write_volatile::<TxmColor>(c, TxmColor { ch: ' ' as u8, bg: GPUBUF_CHAR_BG_BLACK });
            }
        }
    }
}

impl OutputBuffer for TxmBuf {
    fn putc(&mut self, c: char) {
        assert!(self.cury < GPUBUF_SCREEN_HEIGHT);
        assert!(self.curx < GPUBUF_SCREEN_WIDTH);

        unsafe {
            write_volatile::<TxmColor>(&mut self.buf[self.cury][self.curx] as *mut TxmColor, TxmColor { ch: c as u8, bg: 0x0f });
        }

        self.curx += 1;
        if self.curx >= GPUBUF_SCREEN_WIDTH {
            self.curx = 0;
            self.cury += 1;
            if self.cury >= GPUBUF_SCREEN_HEIGHT {
                self.cury = 0;
            }
        }
    }

    fn puts(&mut self, s: &str) {

        self.clear_cur_line();

        for c in s.chars() {
            self.putc(c);
        }

        self.line_feed();
    }

    fn line_feed(&mut self) {
        self.curx = 0;
        self.cury += 1;
        if self.cury >= GPUBUF_SCREEN_HEIGHT {
            self.cury = 0;
        }
    }
}

pub fn make_early_txmbuf() -> TxmBuf {
    TxmBuf {
        buf: unsafe { &mut *(GPUBUF as * mut [[TxmColor; GPUBUF_SCREEN_WIDTH]; GPUBUF_SCREEN_HEIGHT]) },
        curx: 0,
        cury: 3,
    }
}

